### PR TITLE
Fuzzing update

### LINKED ISSUE #109: Fuzzer Update
We want to update the community more regularly about what we're working on. As part of this effort, let's write about what's the lastest news / interesting stuff happening in the fuzzing land. :)

### PR DESCRIPTION
**Pull request by bshastry**
Closes #109

### CODE REVIEW

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+#### Back-end Fuzzers
+
+In order to test if code generation is sound, valid program (input) needs to be fed to the compiler.
+If the program contains syntactic (`iff` instead of `if`) or semantic (use undeclared identifier) errors, it won't be accepted by the compiler.
+Semantic fuzzers are designed to ensure program input does not contain such errors.
+
+To implement semantic fuzzers, we rely on Google's [libprotobuf-mutator][2] library.
+Broadly speaking, writing a semantic fuzzer involves the following steps:
+  - Write a (context-free) grammar for Solidity in the protobuf interface definition language.
+  - Implement a converter interface that accepts a generation of the protobuf grammar and converts into target language
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:97]:
```suggestion
  - Implement a converter interface that accepts a generation of the protobuf grammar and converts into target language.
```

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+                        // This assignment is redundant
+                        x := 42
+                        continue
+                }
+                mstore(0, x)
+        }
+}
+```
+
+The redundant assignment eliminator eliminated all assignments to `x` leading to `mstore(0, x)` thereby optimising the program to one containing `mstore(0, 0)` instead of `mstore(0, 1337)`.
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:129]:
```suggestion
The redundant assignment eliminator eliminated all assignments to `x` leading to `mstore(0, x)`, thereby optimising the program to one containing `mstore(0, 0)` instead of `mstore(0, 1337)`.
```

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+This [bug][8] has been [fixed in `0.6.1`][9].
+
+How does the fuzzer generate such programs?
+The answer, in a nutshell, is the following. In each fuzzing iteration:
+  - libprotobuf-mutator creates a generation of the protobuf grammar for Yul
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:134]: Add ``.``at the end of each bullet point. And optional start with a capital letter.

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  - libprotobuf parses this generation to ensure it is a valid protobuf message
+  - if valid, the protobuf message is accepted by the converter that we have written
+  - the converter converts this protobuf message into a yul program
+
+If you are interested in taking a look at the protobuf grammar for Yul and the converter, please take a look [here][6] and [there][7].
+Contributions for improving the grammar/converter welcome as always.
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:140]:
```suggestion
Contributions for improving the grammar/converter are welcome as always.
```

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+#### Work in Progress
+
+Currently, we are working on a new Solidity program generator that is independent of libprotobuf and libprotobuf-mutator.
+The main reason for this is that the protobuf IDL is hardly suited to capturing programming language semantics.
+Moreover, it is quite laborious to write context-free grammar for high level programming languages.
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:146]:
```suggestion
Moreover, it is quite laborious to write context-free grammar for high-level programming languages.
```

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+The end goal of the new fuzzer generator is to be able to automatically generate a well-formed Solidity program.
+Once this is implemented, we plan to compile/optimise and potentially execute the resulting bytecode on the EVM virtual machine.
+This will help us discover otherwise difficult-to-find bugs introduced by the compiler that show up only when EVM state is examined.
+Watch this space for more updates on Solidity fuzzing!
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:151]:
```suggestion

Watch this space for more updates on Solidity fuzzing!
```

Comment (franzihei) [reply _posts/2021-02-05-solidity-fuzzer-update.md:151]: (add line break)

**Review by franzihei — COMMENTED** [review]
Very cool and informative article! I've added a few comments for improvements, but nothing major. :)

File: _posts/2021-02-05-solidity-fuzzer-update.md (hunk 3 above)
Comment (bshastry) [reply _posts/2021-02-05-solidity-fuzzer-update.md]: Done

**Comment by bshastry** [comment]
Thank you for your review @franzihei. I've accepted all your suggestions and squashed everything into a single commit.

**Comment by franzihei** [comment]
Great @bshastry! I think there are a few remaining comments from my side still open though. (See the unresolved ones in the files page.)

**Comment by bshastry** [comment]
> Great @bshastry! I think there are a few remaining comments from my side still open though. (See the unresolved ones in the files page.)

Oops, did a second pass, hope I covered all of them now :-)

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+Watch this space for more updates on Solidity fuzzing!
+
+[1]: https://github.com/google/oss-fuzz
+[2]: https://github.com/google/libprotobuf-mutator
+[3]: https://docs.soliditylang.org/en/develop/natspec-format.html
+[4]: https://github.com/ethereum/solidity/issues/10433
+[5]: https://raw.githubusercontent.com/ethereum/solidity/develop/test/tools/ossfuzz/config/solidity.dict
+[6]: https://raw.githubusercontent.com/ethereum/solidity/develop/test/tools/ossfuzz/yulProto.proto
+[7]: https://raw.githubusercontent.com/ethereum/solidity/develop/test/tools/ossfuzz/protoToYul.cpp
```
Comment (hrkrshnn) [inline _posts/2021-02-05-solidity-fuzzer-update.md:160]:
```suggestion
[7]: https://github.com/ethereum/solidity/blob/develop/test/tools/ossfuzz/protoToYul.cpp
```
Similarly for the other `raw.github...` links

**Review by hrkrshnn — COMMENTED** [review]
Minor comment about links, since the regular github link looks better.

Overall, it's a nice article :)

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+layout: post
+published: true
+title: 'Solidity Fuzzer Update'
+date: '2021-02-05'
+author: Bhargava Shastry
+category: Explainers
+---
+
+Security vulnerabilities and bugs detract from software quality.
+To discover them early, we have adopted fuzz testing: feeding randomly generated programs to the Solidity compiler and observing the compilation runtime and code generated.
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:11]:
```suggestion
To discover them early, we have adopted fuzz testing: feeding randomly generated programs to the Solidity compiler and observing the compilation runtime and code generated.

```

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+Since Q1 2019, the Solidity compiler is fuzz tested via Google's [open-source software fuzz][1] (oss-fuzz) framework.
+In this post, we briefly describe the work that has been done on this front, and work that is currently in progress.
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:13]:
```suggestion

In this post, we briefly describe the work that has been done on this front, and work that is currently in progress.
```

File: _posts/2021-02-05-solidity-fuzzer-update.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+### Fuzzing Overview
+
+Broadly speaking, we have developed two classes of fuzzers:
+  - Front-end fuzzers that test the compiler front-end (parser/analyser).
+  - Back-end fuzzers that test the compiler back-end (optimiser).
+
```
Comment (franzihei) [inline _posts/2021-02-05-solidity-fuzzer-update.md:20]: Here the spacing looks super tight in the preview but I don't know if there's a way we can fix it.

**Review by franzihei — COMMENTED** [review]
Just 3 more tiny things with regards to formatting. (You can see how the post will look like here: https://deploy-preview-124--solidity-blog.netlify.app/2021/02/05/solidity-fuzzer-update/)

File: _posts/2021-02-05-solidity-fuzzer-update.md (hunk 10 above)
Comment (bshastry) [reply _posts/2021-02-05-solidity-fuzzer-update.md:22]: Ah right, thanks for the preview link. I will try to reformat the text a little bit more so it looks a little more tidy. I suppose one thing that can be done is to increase heading size since we don't have numbered headers i.e., `Section 1` and `Section 1.1` can have the same heading size making them look better formatted although `1.1` is "below" `1`

**Comment by franzihei** [comment]
Wrt the title we could keep the current one (``Solidity Fuzzer Update``) or, alternatively, could use something like:
- Solidity Fuzzer Introduction & Update
- How Solidity uses Fuzz Testing
- Fuzzing the Solidity Compiler
- An Intro to Solidity's Fuzz Testing Efforts

Trying to make it a bit more catchy, wdyt?

**Review by franzihei — APPROVED** [review]
Last remaining things from my side:
- Decide on keeping title or changing it
- Updating the date before publishing :)

**Comment by bshastry** [comment]
> Wrt the title we could keep the current one (`Solidity Fuzzer Update`) or, alternatively, could use something like:
> 
> * Solidity Fuzzer Introduction & Update
> * How Solidity uses Fuzz Testing
> * Fuzzing the Solidity Compiler
> * An Intro to Solidity's Fuzz Testing Efforts
> 
> Trying to make it a bit more catchy, wdyt?

Yeah, perhaps `Solidity Fuzzer Update` is a bit odd since it is the first ever (iiuc) fuzzing post on the Solidity blog. I like your suggestions, particularly `An Intro to Solidity's...` and `How Solidity uses Fuzz Testing.`

**Comment by bshastry** [comment]
> * Updating the date before publishing :)

Should I rename the post with the date on which the post will go public? Or is it sufficient to edit the `date` header in the post text?

**Comment by franzihei** [comment]
Date: I always rename both the post and change the info in the header. The relevant one is the header but for archive purposes it makes sense to have the file names match the publish dates. It's already late today and Chris didn't review yet so let's aim to publish this tomorrow.

Title: I also like ``An Intro to Solidity's... `` - maybe let's use ``introduction`` instead of ``intro`` and (optional) ``approach``instead of ``efforts`` --> ``Introduction to Solidity's Fuzz Testing Approach``? I don't have strong preferences as long as we make it more specific than just "update". You choose. :)

**Comment by bshastry** [comment]
> Date: I always rename both the post and change the info in the header. The relevant one is the header but for archive purposes it makes sense to have the file names match the publish dates. It's already late today and @chriseth didn't review yet so let's aim to publish this tomorrow.
> 
> Title: I also like `An Intro to Solidity's... ` - maybe let's use `introduction` instead of `intro` and (optional) `approach`instead of `efforts` --> `Introduction to Solidity's Fuzz Testing Approach`? I don't have strong preferences as long as we make it more specific than just "update". You choose. :)

Cool, thank you once again for your detailed review. I'll update the title and date (make it tomorrow). Can update accordingly just before publishing if necessary.

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+title: 'An Introduction to Solidity's Fuzz Testing Approach'
+date: '2021-02-09'
+author: Bhargava Shastry
+category: Explainers
+---
+
+Security vulnerabilities and bugs detract from software quality.
+To discover them early, we have adopted fuzz testing: feeding randomly generated programs to the Solidity compiler and observing the compilation runtime and code generated.
+
+Since Q1 2019, the Solidity compiler is fuzz tested via Google's [open-source software fuzz][1] (oss-fuzz) framework.
```
Comment (chriseth) [inline _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:13]: Does the link render properly with the space?

Comment (chriseth) [reply _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:13]: ah, it's just a regular space, not part of the link, right?

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+layout: post
+published: true
+title: 'An Introduction to Solidity's Fuzz Testing Approach'
+date: '2021-02-09'
+author: Bhargava Shastry
+category: Explainers
+---
+
+Security vulnerabilities and bugs detract from software quality.
+To discover them early, we have adopted fuzz testing: feeding randomly generated programs to the Solidity compiler and observing the compilation runtime and code generated.
```
Comment (chriseth) [inline _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:11]:
```suggestion
To discover them early, at best before they are released, we have adopted fuzz testing: feeding randomly generated programs to the Solidity compiler and observing the compilation runtime and code generated.
```

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+}
+```
+
+The mutations prepended the `///@return` line to the modifier declaration and copied the modifier declaration to the derived contract.
+
+A natural question is the following: How did the fuzzer "know" that this particular mutation will lead to an error?
+The answer is that it does not know; the fuzzer simply mutates existing input to create new inputs and it just so happened that this mutation discovered a bug.
+A follow-up question could be: How are mutations created?
+Broadly speaking, mutations fall into one of the following categories:
+  - Flipping bit(s) at certain position(s) in the input stream.
```
Comment (chriseth) [inline _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:60]: Does the singular make sense for these? It at least should be `a bit`, so maybe just stick with `bits`?

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+---
+
+Security vulnerabilities and bugs detract from software quality.
+To discover them early, we have adopted fuzz testing: feeding randomly generated programs to the Solidity compiler and observing the compilation runtime and code generated.
+
+Since Q1 2019, the Solidity compiler is fuzz tested via Google's [open-source software fuzz][1] (oss-fuzz) framework.
+
+In this post, we briefly describe the work that has been done on this front, and work that is currently in progress.
+
+### Fuzzer Overview
```
Comment (chriseth) [inline _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:17]: Not sure where to put that best, but maybe mention that the compiler makes very heavy use of assertions which are then hit by the fuzzer?

Comment (chriseth) [reply _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:17]: Actually you already mention that "led to an assertion failure" - maybe just extend on that?
Perhaps also mention that bugs in the frontend almost never lead to invalid code because of that.

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+
+**Notes**:
+
+This [issue][4] is a few months old and has been fixed in `0.7.6`.
+The issue is that Solidity modifiers don't have return variables, and hence a `@return` NatSpec tag does not make sense.
+The program above should have resulted in a parsing error (which is what the fix ensures) but instead resulted in a failed assertion.
+
+### Back-end Fuzzers
+
+In order to test if code generation is sound, valid program (input) needs to be fed to the compiler.
```
Comment (chriseth) [inline _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:92]:
```suggestion
In order to test if code generation is sound, a valid program (input) needs to be fed to the compiler.
```

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+```
+{
+        for {let i:= 0} lt(i,2) {i := add(i,1)}
+        {
+                // x is declared and implicitly
+                // initialized to zero.
+                let x
+                // This assignment is not redundant
+                // since it is used by mstore statement
+                // that follows the if statement
```
Comment (chriseth) [inline _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:119]:
```suggestion
                // since it is used by the mstore statement
                // after the if statement
```

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 17]
```diff
@@ -0,0 +1,164 @@
...
+                x := 1337
+                if lt(i,1) {
+                        // This assignment is redundant
```
Comment (chriseth) [inline _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:122]:
```suggestion
                        // This assignment is redundant because of the `continue`
```

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md (hunk 11 above)
Comment (bshastry) [reply _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:13]: Yeah, the space is part of the hyperlinked text iiuc

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md (hunk 13 above)
Comment (bshastry) [reply _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md]: Done

File: _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md (hunk 14 above)
Comment (bshastry) [reply _posts/2021-02-09-an-introduction-to-soliditys-fuzz-testing-approach.md:17]: Done.

**Comment by bshastry** [comment]
Only thing left to do is to acknowledge external contributors who reported bugs. Any ideas how to word it @chriseth ?

I was thinking of something along the following lines

```
Something like "We'd like to think Alex Groce for independently fuzzing the Solidity compiler and reporting bugs."
```

**Comment by chriseth** [comment]
Sounds good! Only typo: think -> thank :)

**Comment by bshastry** [comment]
Acknowledged external contributors and updated date of blog post to today.

File: _posts/2021-02-10-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+Broadly speaking, we have developed two classes of fuzzers:
+  - Front-end fuzzers that test the compiler front-end (parser/analyser).
+  - Back-end fuzzers that test the compiler back-end (optimiser).
+
+### Front-end Fuzzers
+
+We have developed front-end fuzzers in order to test if programs are parsed and analysed correctly.
+Front-end fuzzing comprises mutating existing test cases (most often unit tests) in order to test how parsing and program analysis respond to corner case input.
+
+The Solidity compiler makes heavy use of assertions during the input program parsing/analysis phase.
```
Comment (chriseth) [inline _posts/2021-02-10-an-introduction-to-soliditys-fuzz-testing-approach.md:28]: Actually mainly during code generation...

File: _posts/2021-02-10-an-introduction-to-soliditys-fuzz-testing-approach.md [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+Inputs generated by front-end fuzzers may fail such assertions, signalling an error in the compiler front-end.
+Please note that bugs that involve failed compiler assertions almost never lead to invalid code generation because code generation is initiated only when the parsing/analysis phase is successful.
```
Comment (chriseth) [inline _posts/2021-02-10-an-introduction-to-soliditys-fuzz-testing-approach.md:30]:
```suggestion
Fuzzing is effective because the Solidity compiler makes heavy use of assertions both during the input program parsing/analysis phase as well as the code generation phase.
Inputs generated by front-end fuzzers may fail such assertions, signalling an error in the compiler.
```
(Yeah, maybe we don't have to mention it that explicitly)

**Comment by franzihei** [comment]
There seems to be an issue with the formatting now (title is gone in the preview). 
image

**Comment by bshastry** [comment]
> There seems to be an issue with the formatting now (title is gone in the preview).
> image

I can reproduce this locally as well. The error seems to indicate a problem with `yaml` header in the post but I simply copied these fields from another post. So I don't understand what is happening.

**Comment by bshastry** [comment]
Ah it's a problem with the use of apostrophe in the title :( Will try to fix

**Comment by bshastry** [comment]
I pushed a fix by un-single-quoting the title but retaining the apostrophe in the title. To be pedantic, we need to rename the post to contain the apostrophe as well but I'm not sure this is a good idea (i.e., to have `solidity's-*.md` as the file name).

**Comment by franzihei** [comment]
Looks good for me now!

**Review by chriseth — APPROVED** [review]
